Improve handling of common Regex sets

stephentoub opened the pull request:
This improves how many of the most common character classes are emitted.

- Today for a set like [\r\n], we'll emit a comparison that compares the char to each of '\r' and '\n', but for a set like [^\r\n], we end up falling back to emitting a lookup table.  With this PR, we simply use the existing support for the non-negating case, just negating the result.
- Today for a set like [\p{IsGreek}\p{IsGreekExtended}] that ends up being two ranges, we'll fall back to our lookup table.  With this PR, we'll emit it as two range checks.
- Today for a set like [A-Za-z], we'll fall back to our lookup table.  As a special case of two-range support, with this PR we'll now recognize that these ranges are just one bit flip away from each other, and we'll employ the normal ASCII casing to do a single range comparison against the input or'd with a mask.
- Today as a fallback, we employ a lookup table stored in a string; this requires a bounds check, dereferencing the string object, doing the math to find the right index, doing the math to find the right bit, etc.  With this PR, for sets composed only of ranges where the exclusiveMax - inclusiveMin <= 64, with this PR we'll now emit it as a lookup into a ulong that's done in a branchless fashion and is much faster.
- It appears to be relatively common for folks to use [\d\D], [\w\W], or [\s\S] as a simple way of saying "match anything"; RegexOptions.Singleline changes '.' to mean this as well.  We already have special handling for '.' with Singleline as "AnyClass"... this just normalizes those other common representations into the same shape so that everyhing else recognizes them accordingly.
- Today when we see an AnyClass, we emit a nonsense comparison that always results in true (or false for negations); that's because, for a while, the expression given to the matching routine may have had side effects.  There are no longer side effects, though, so it's ok to just emit "true" or "false" directly and make the operation cheaper.

Fixes https://github.com/dotnet/runtime/issues/67056
Fixes https://github.com/dotnet/runtime/issues/62647
@joperezr, this will have conflicts with your https://github.com/dotnet/runtime/pull/67184.  I'll wait until that goes in to merge this.

danmoseley commented on src/libraries/System.Text.RegularExpressions/gen/RegexGenerator.Emitter.cs:
```diff
@@ -3908,13 +3910,14 @@ private static string MatchCharacterClass(bool hasTextInfo, RegexOptions options
             // the very common case with case insensitivity of two characters next to each other being the upper and lowercase
             // ASCII variants of each other, in which case we can use bit manipulation to avoid a comparison.
-            if (!invariant && !RegexCharClass.IsNegated(charClass))
+            if (!invariant && (!caseInsensitive || !RegexCharClass.IsNegated(charClass)))
```
```suggestion
            if (!invariant && !(caseInsensitive && RegexCharClass.IsNegated(charClass)))
```
clearer?

stephentoub replied:
The caseInsensitive piece is temporary and will go away after https://github.com/dotnet/runtime/pull/67184.

danmoseley commented on src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexNode.cs:
```diff
@@ -889,6 +889,20 @@ private RegexNode ReduceSet()
+            switch (Str)
+            {
+                // Different ways of saying "match anything"
+                case RegexCharClass.WordNotWordClass:
```
we mostly don't have tests for `[\d\D]` specifically - worth adding a test for each? that would verify that they still have correct behavior.

stephentoub replied:
Added

danmoseley commented:
What is our bar for adding optimizations to the interpreter as well?

joperezr commented on src/libraries/System.Text.RegularExpressions/gen/RegexGenerator.Emitter.cs:
```diff
@@ -3908,13 +3910,14 @@ private static string MatchCharacterClass(bool hasTextInfo, RegexOptions options
             // the very common case with case insensitivity of two characters next to each other being the upper and lowercase
             // ASCII variants of each other, in which case we can use bit manipulation to avoid a comparison.
-            if (!invariant && !RegexCharClass.IsNegated(charClass))
+            if (!invariant && (!caseInsensitive || !RegexCharClass.IsNegated(charClass)))
```
Would it be worth it to call RegexCharClass.GetSetChars before the ifcheck in order to be able to add to the if clause an additional `... && setChars.Count <= 3)` ?

joperezr commented on src/libraries/System.Text.RegularExpressions/gen/RegexGenerator.Emitter.cs:
```diff
@@ -3936,21 +3940,87 @@ private static string MatchCharacterClass(bool hasTextInfo, RegexOptions options
+                RegexCharClass.TryGetDoubleRange(charClass, out (char LowInclusive, char HighInclusive) rangeUnmasked, out (char LowInclusive, char HighInclusive) rangeMasked) &&
+                rangeUnmasked.LowInclusive != rangeUnmasked.HighInclusive && // more than a single char in range0 (just to simplify, since single chars should have already been handled)
+                rangeMasked.LowInclusive != rangeMasked.HighInclusive && // more than a single char in range1 (just to simplify, since single chars should have already been handled)
+                RegexCharClass.DifferByOneBit(rangeUnmasked.LowInclusive, rangeMasked.LowInclusive, out int lowMask) &&
```
Is there no case where the Low and high inclusive ends of the ranges might have the same mask but some character in the middle of the range have a different mask? I guess my question is if it is safe to just check that the masks of the beginning and end of the ranges and assume all characters in between have the same mask.

stephentoub replied:
It's making my head hurt to try to answer that, so since really the only thing that matters are ASCII letters, I've just narrowed this down to handling that.

joperezr replied:
FWIW, I *think* your assumption seemed right, but I believe it wouldn’t be trivial to prove this so my question was mostly me wondering whether you knew this to be true.

joperezr commented on src/libraries/System.Text.RegularExpressions/gen/RegexGenerator.Emitter.cs:
```diff
@@ -3936,21 +3940,87 @@ private static string MatchCharacterClass(bool hasTextInfo, RegexOptions options
+            // Next, handle sets where the high - low + 1 range is <= 64.  In that case, we can emit
+            // a branchless lookup in a ulong that does not rely on loading any objects (e.g. the string-based
+            // lookup we use later).  This nicely handles common sets like [0-9A-Fa-f], [0-9a-f], [A-Za-z], etc.
+            if (!caseInsensitive &&
```
Should this also check if !invariant?

stephentoub replied:
No, invariant is only relevant if caseInsensitive is true.

stephentoub committed:
Improve handling of more common Regex sets

- Today for a set like [\r\n], we'll emit a comparison that compares the char to each of '\r' and '\n', but for a set like [^\r\n], we end up falling back to emitting a lookup table.  With this PR, we simply use the existing support for the non-negating case, just negating the result.
- Today for a set like [\p{IsGreek}\p{IsGreekExtended}] that ends up being two ranges, we'll fall back to our lookup table.  With this PR, we'll emit it as two range checks.
- Today for a set like [A-Za-z], we'll fall back to our lookup table.  As a special case of two-range support, with this PR we'll now recognize that these ranges are just one bit flip away from each other, and we'll employ the normal ASCII casing to do a single range comparison against the input or'd with a mask.
- Today as a fallback, we employ a lookup table stored in a string; this requires a bounds check, dereferencing the string object, doing the math to find the right index, doing the math to find the right bit, etc.  With this PR, for sets composed only of ranges where the exclusiveMax - inclusiveMin <= 64, with this PR we'll now emit it as a lookup into a ulong that's done in a branchless fashion and is much faster.
- It appears to be relatively common for folks to use [\d\D], [\w\W], or [\s\S] as a simple way of saying "match anything"; RegexOptions.Singleline changes '.' to mean this as well.  We already have special handling for '.' with Singleline as "AnyClass"... this just normalizes those other common representations into the same shape so that everyhing else recognizes them accordingly.
- Today when we see an AnyClass, we emit a nonsense comparison that always results in true (or false for negations); that's because, for a while, the expression given to the matching routine may have had side effects.  There are no longer side effects, though, so it's ok to just emit "true" or "false" directly and make the operation cheaper.
- For every optimization we have in MatchCharacterClass, we should always be able to handle negation trivially.
- Handle character classes composed of multiple UnicodeCategories. This helps with composed categories, like \p{N}.
- Fix hard-coded char class strings for \W and \S. There are multiple ways to invert a RegexCharClass string: you can invert the whole string by just setting the invert flag, or you can invert all the individual components, e.g. if the string is composed of only categories, invert each category.  The hardcoded string the parser uses when you write \W simply sets the negated bit, but this causes problems if \W is used as [\W], because then the individual components are added into a larger set that doesn't have negation set.  And that means \W and [\W] result in different strings, which means any place we special-case the string for \W, we don't recognize [\W].  The same applies to \S. This commit changes the hardcoded string for \W and \S to use the more canonical form. Also, the implementation generally uses "set" and "class" interchangeably, but when specifying the ECMA-related strings, it uses "set" to actually mean "ranges", which is very confusing.  I've changed them.
